ethers-core/ethers-derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+//!
+//! The trait used to derive the macro is found in `ethers_core::transaction::eip712::Eip712`
+//! Both the derive macro and the trait must be in context when using
+//!
+//! This derive macro requires the `#[eip712]` attributes to be included
+//! for specifying the domain separator used in encoding the hash.
+//!
+//! NOTE: In addition to deriving `Eip712` trait, the `EthAbiType` trait must also be derived.
+//! This allows the struct to be parsed into `ethers_core::abi::Token` for encoding.
+//!
+//! # Optional Eip712 Parameters
+//!
+//! The only optional parameter is `salt`, which accepts a string
+//! that is hashed using keccak256 and stored as bytes.
+//!
+//! # Example Usage
+//!
+//! ```rust
+//! use ethers_contract::EthAbiType;
+//! use ethers_derive_eip712::*;
+//! use ethers_core::types::{transaction::eip712::Eip712, H160};
+//!
+//! #[derive(Debug, Eip712, EthAbiType)]
+//! #[eip712(
+//!     name = "Radicle",
+//!     version = "1",
+//!     chain_id = 1,
+//!     verifying_contract = "0x0000000000000000000000000000000000000000"
+//!     // salt is an optional parameter
+//!     salt = "my-unique-spice"
+//! )]
+//! pub struct Puzzle {
+//!     pub organization: H160,
+//!     pub contributor: H160,
+//!     pub commit: String,
+//!     pub project: String,
+//! }
+//!
+//! let puzzle = Puzzle {
+//!     organization: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     contributor: "0000000000000000000000000000000000000000"
+//!         .parse::<H160>()
+//!         .expect("failed to parse address"),
+//!     commit: "5693b7019eb3e4487a81273c6f5e1832d77acb53".to_string(),
+//!     project: "radicle-reward".to_string(),
+//! };
+//!
+//! let hash = puzzle.encode_eip712()?;
+//! ```
+//!
+//! # Limitations
+//!
+//! At the moment, the derive macro does not recursively encode nested Eip712 structs.
+//!
+//! There is an Inner helper attribute `#[eip712]` for fields that will eventually be used to
+//! determine if there is a nested eip712 struct. However, this work is not yet complete.
+//!
+use std::convert::TryFrom;
+
+use ethers_core::types::transaction::eip712;
+use proc_macro::TokenStream;
+use quote::quote;
+
+#[proc_macro_derive(Eip712, attributes(eip712))]
+pub fn eip_712_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("failed to parse token stream for Eip712 derived struct");
+
+    impl_eip_712_macro(&ast)
+}
+
+// Main implementation macro, used to compute static values and define
+// method for encoding the final eip712 payload;
+fn impl_eip_712_macro(ast: &syn::DeriveInput) -> TokenStream {
+    // Primary type should match the type in the ethereum verifying contract;
+    let primary_type = &ast.ident;
+
+    // Computer domain separator
+    let domain = match eip712::EIP712Domain::try_from(ast) {
+        Ok(attributes) => attributes,
+        Err(e) => return TokenStream::from(e),
```

Comment:
`syn::Error`

Reply:
I think it is okay to do this here. This is returning a syn::Error, however, we don't have access to a span on the `DeriveInput` input. I see this is also how `ethers-contract-derive` is handling it, [here](https://github.com/gakonst/ethers-rs/blob/master/ethers-contract/ethers-contract-derive/src/lib.rs#L128).

Open to exploring directly using `syn::Error`, but not sure what the idiomatic way is to handle the error without the span, and changing the `proc_macro::TokenStream` to `proc_macro2::TokenStream`.

Reply:
yeah you're right, sgtm